File: 03_Greedy/job_scheduling.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
	size_t id ; 
	size_t deadline ; 
	size_t profit ; 
} Job ; 

size_t greedy_scheduling(
	Job* jobs , 
	size_t n
) {
    // Sort the jobs in the decreasing order of their profit
	for( int i = 0 ; i < n ; i++ ) {
		int j = i ; 
		while( jobs[j].profit > jobs[j-1].profit && j > 0 ) {	
			Job temp = jobs[j-1] ; 
			jobs[j-1] = jobs[j] ; 
			jobs[j] = temp ;
			j-- ; 
		}
	}

    // Calculate max. permissible time to
    // complete the jobs
	size_t max_deadline = 0;
	for( size_t i = 0 ; i < n ; i++ ) {
		if( jobs[i].deadline > max_deadline ) {
			max_deadline = jobs[i].deadline ; 
		}
	}

    // Select jobs from the sorted jobs array
    // respecting the max_deadline
	size_t remaining_time = max_deadline ; 
	size_t profit = 0 ; 
	for( size_t i = 0 ; i < n ; i++ ) {
		if( jobs[i].deadline <= remaining_time ) {
			remaining_time -= jobs[i].deadline ;
			profit += jobs[i].profit ;  
		}
	} 
	
    return profit ; 
}

int main( int argc , char** argv ) {
	Job jobs[] = {
		{ 1 , 2 , 20 } , 
		{ 2 , 2 , 60 } , 
		{ 3 , 1 , 40 } , 
		{ 4 , 3 , 100 } , 
		{ 5 , 4 , 80 }
	} ; 
    size_t max_profit = greedy_scheduling( jobs , 5 ) ;
    printf( "Maximum profit is %li" , max_profit  ) ;  
	return 0;
}

File: 03_Greedy/job_scheduling.py
class Job:
    def __init__(self, id: int, profit: int, deadline: int):
        self.id = id
        self.deadline = deadline
        self.profit = profit

    def __str__(self) -> str:
        return f"Job: id={self.id}, deadline={self.deadline}, profit={self.profit}"


def greedy_scheduling(jobs: list[Job]):
    # Sort the jobs in decreasing order of their profits
    for i in range(len(jobs)):
        j = i
        while jobs[j].profit > jobs[j - 1].profit and j > 0:
            temp = jobs[j - 1]
            jobs[j - 1] = jobs[j]
            jobs[j] = temp
            j -= 1

    # Calculate max permissible time to complete the jobs
    # Reference: https://youtu.be/zPtI8q9gvX8?si=8mHVmh4-q4T_d234
    max_deadline = max([job.deadline for job in jobs])
    slot_status = [True for _ in range(max_deadline)]
    max_profit_seq = []
    for job in jobs:
        for i in range(job.deadline - 1, -1, -1):
            if slot_status[i]:
                slot_status[i] = False
                max_profit_seq.append(job)
                break
    for job in max_profit_seq:
        print(job)

    max_profit = sum([job.profit for job in max_profit_seq])
    return max_profit


jobs = [Job(1, 20, 2), Job(2, 15, 2), Job(3, 10, 1), Job(4, 5, 3), Job(5, 1, 3)]
max_profit = greedy_scheduling(jobs)
print(f"Max profit is {max_profit}")
